File: leetcode/solutions/0543.DiameterOfBinaryTree/diameterOfBinaryTree.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";

export default function diameterOfBinaryTree(
  root: BinaryTreeNode<number> | null
): number {
  if (!root || (!root.left && !root.right)) {
    return 0;
  }

  const [maxLen, maxSum] = dfs(root);

  return Math.max(maxLen - 1, maxSum);
}

function dfs(node: BinaryTreeNode<number> | null): [number, number] {
  if (!node) {
    return [0, 0];
  }

  const [len1, sum1] = dfs(node.left);
  const [len2, sum2] = dfs(node.right);

  return [Math.max(len1, len2) + 1, Math.max(len1 + len2, sum1, sum2)];
}

File: leetcode/solutions/0543.DiameterOfBinaryTree/diameterOfBinaryTree.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// TreeNode structure for LeetCode
type TreeNode = datastructure.TreeNode

func diameterOfBinaryTree(root *TreeNode) int {
	if root == nil || root.Left == nil && root.Right == nil {
		return 0
	}

	maxLen, maxSum := dfs(root)

	return max(maxLen-1, maxSum)
}

func dfs(node *TreeNode) (int, int) {
	if node == nil {
		return 0, 0
	}

	len1, sum1 := dfs(node.Left)
	len2, sum2 := dfs(node.Right)

	return max(len1, len2) + 1, max(len1+len2, sum1, sum2)

}

func max(nums ...int) int {
	max := nums[0]

	for i := 1; i < len(nums); i++ {
		if max < nums[i] {
			max = nums[i]
		}
	}

	return max
}
